File: src/main/java/fi/quanfoxes/parser/Resolver.java
package fi.quanfoxes.parser;

import java.util.ArrayList;
import java.util.List;

import fi.quanfoxes.Types;
import fi.quanfoxes.lexer.Operators;
import fi.quanfoxes.parser.nodes.FunctionNode;
import fi.quanfoxes.parser.nodes.OperatorNode;
import fi.quanfoxes.parser.nodes.TypeNode;
import fi.quanfoxes.types.Number;

public class Resolver {
    
    /**
    * Tries to resolve any unresolved nodes in a node tree
    * @param context Context to use when resolving
    * @param node Node tree
    * @param errors Output list for errors
    * @return Returns a resolved node tree on success, otherwise null
    */
    public static Node resolve(Context context, Node node, List<Exception> errors) {
        if (node instanceof Resolvable) {
            Resolvable resolvable = (Resolvable)node;

            try {
                Node resolved = resolvable.resolve(context);
                return resolved == null ? node : resolved;
            }
            catch (Exception e) {
                errors.add(e);
            }

            return null;
        }
        else {
            Node iterator = node.first();

            while (iterator != null) {
                Node resolved;

                if (iterator instanceof TypeNode) {
                    TypeNode type = (TypeNode)iterator;
                    resolved = Resolver.resolve(type.getType(), iterator, errors);
                    Resolver.resolveVariables(type.getType(), errors);
                }
                else if (iterator instanceof FunctionNode) {
                    FunctionNode function = (FunctionNode)iterator;
                    resolved = Resolver.resolve(function.getFunction(), iterator, errors);
                    Resolver.resolveVariables(function.getFunction(), errors);
                }
                else {
                    resolved = Resolver.resolve(context, iterator, errors);
                }

                if (resolved != null) {
                    iterator.replace(resolved);
                }

                iterator = iterator.next();
            }

            return node;
        }
    }

    private static Type getSharedNumber(Number a, Number b) {
        return a.getBitCount() > b.getBitCount() ? a : b;
    }

    public static Context getSharedContext(Context a, Context b) {
        if (a instanceof Number && b instanceof Number) {
            return getSharedNumber((Number)a, (Number)b);
        }

        Context context = a;
        
        while (context != null) {
            Context iterator = b;

            while (iterator != null) {
                
                if (iterator == context) {
                    return context;
                }
                
                iterator = iterator.getParent();
            }

            context = context.getParent();
        }

        return null;
    }

    /**
     * Returns the shared type between the given types
     * @return Success: Shared type between the given types, Failure: null
     */
    public static Type getSharedType(Type a, Type b) {
        if (a == b) {
            return a;
        }

        if (a instanceof Number && b instanceof Number) {
            return Resolver.getSharedNumber((Number)a, (Number)b);
        }

        for (Type type : a.getSuperTypes()) {
            if (b.getSuperTypes().contains(type)) {
                return type;
            }
        }

        return Types.UNKNOWN;
    }

    /**
     * Returns the shared type between the given types
     * @param types List of types to solve
     * @return Success: Shared type between the given types, Failure: null
     */
    public static Type getSharedType(List<Type> types) {
        if (types.isEmpty()) {
            return Types.UNKNOWN;
        }
        else if (types.size() == 1) {
            return types.get(0);
        }

        Type current = types.get(0);

        for (int i = 1; i < types.size(); i++) {
            if (current == null) {
                break;
            }

            current = Resolver.getSharedType(current, types.get(i));
        }

        return current;
    }

    /**
     * Returns all child node types
     * @return Types of children nodes
     */
    public static List<Type> getTypes(Node node) {
        List<Type> types = new ArrayList<>();
        Node iterator = node.first();
        
        while (iterator != null) {
            if (iterator instanceof Contextable) {
                Contextable contextable = (Contextable)iterator;
                Context context = null;

                try {
                    context = contextable.getContext();
                }
                catch (Exception e) {
                    return null;
                }

                if (context == null || !context.isType()) {
                    return null;
                }
                else {
                    types.add((Type)context);
                }
            }

            iterator = iterator.next();
        }

        return types;
    }

    public static void resolveVariable(Variable variable) throws Exception {
        List<Type> types = new ArrayList<>();

        for (Node reference : variable.getUsages()) {
            Node parent = reference.parent();

            if (parent != null) {
                if (parent instanceof OperatorNode) {
                    OperatorNode operator = (OperatorNode)parent;

                    // Try to resolve type via contextable right side of the assign operator
                    if (operator.getOperator() == Operators.ASSIGN && 
                            operator.getRight() instanceof Contextable) {
                        
                        try {
                            Contextable contextable = (Contextable)operator.getRight();
                            Context context = contextable.getContext();
                            
                            // Verify the type is resolved
                            if (context != Types.UNKNOWN && context.isType()) {
                                types.add((Type)context);
                            }
                        }
                        catch(Exception e) {
                            continue;
                        }
                    }
                }
            }
        }

        Type shared = Resolver.getSharedType(types);

        if (shared != Types.UNKNOWN) {
            variable.setType(shared);
            return;
        }

        throw new Exception(String.format("Couldn't resolve type of variable '%s'", variable.getName()));
    }

    public static void resolveVariables(Context context, List<Exception> errors) {
        for (Variable variable : context.getVariables()) {
            if (variable.getType() == Types.UNKNOWN) {
                try {
                    resolveVariable(variable);
                }
                catch(Exception e) {
                    errors.add(e);
                }
            }
        }
    }
}

File: Zigzag/Parser/Resolver.cs
using System;
using System.Collections.Generic;

public class Resolver
{
	/// <summary>
	/// Tries to resolve the given node tree
	/// </summary>
	public static void Resolve(Context context, Node node)
	{
		var resolved = ResolveTree(context, node);

		if (resolved != null)
		{
			node.Replace(resolved);
		}
	}

	/// <summary>
	/// Tries to resolve the problems in the given context
	/// </summary>
	public static void ResolveContext(Context context)
	{
		ResolveVariables(context);
		
		foreach (var type in context.Types.Values)
		{
			ResolveVariables(type);
			ResolveContext(type);
		}

		foreach (var implementation in context.GetImplementedFunctions())
		{
			ResolveVariables(implementation);
			ResolveTree(implementation, implementation.Node!);
		}
	}

	/// <summary>
	/// Tries to resolve the problems in the node tree
	/// </summary>
	public static Node? ResolveTree(Context context, Node node)
	{
		if (node is IResolvable resolvable)
		{
			try
			{
				return resolvable.Resolve(context) ?? node;
			}
			catch
			{
				Console.WriteLine("Warning: Resolvable threw an exception");
			}

			return null;
		}
		else
		{
			var iterator = node.First;

			while (iterator != null)
			{
				var resolved = Resolver.ResolveTree(context, iterator);

				if (resolved != null)
				{
					iterator.Replace(resolved);
				}

				iterator = iterator.Next;
			}

			return node;
		}
	}

	/// <summary>
	/// Returns the number type that should be the outcome when using the two given numbers together
	/// </summary>
	private static Type GetSharedNumber(Number a, Number b)
	{
		return a.Bits > b.Bits ? a : b;
	}

	/// <summary>
	/// Returns the shared type between the types
	/// </summary>
	/// <returns>Success: Shared type between the types, Failure: null</returns>
	public static Type? GetSharedType(Type? a, Type? b)
	{
		if (a == b)
		{
			return a;
		}
		else if (a == Types.UNKNOWN || b == Types.UNKNOWN)
		{
			return Types.UNKNOWN;
		}

		if (a is Number && b is Number)
		{
			if (a is Decimal || b is Decimal)
			{
				return Types.DECIMAL;
			}

			return GetSharedNumber((Number)a, (Number)b);
		}

		foreach (Type type in a.Supertypes)
		{
			if (b.Supertypes.Contains(type))
			{
				return type;
			}
		}

		return Types.UNKNOWN;
	}

	 /// <summary>
	 /// Returns the shared type between the types
	 /// </summary>
	 /// <param name="types">Type list to go through</param>
	 /// <returns>Success: Shared type between the types, Failure: null</returns>
	private static Type? GetSharedType(List<Type> types)
	{
		if (types.Count == 0)
		{
			return Types.UNKNOWN;
		}
		else if (types.Count == 1)
		{
			return types[0];
		}

		var current = (Type?)types[0];

		for (var i = 1; i < types.Count; i++)
		{
			if (current == null)
			{
				break;
			}

			current = Resolver.GetSharedType(current, types[i]);
		}

		return current;
	}

	/// <summary>
	/// Returns the types of the child nodes of the given node
	/// </summary>
	/// <returns>Success: Types of the child nodes, Failure: null</returns>
	public static List<Type>? GetTypes(Node node)
	{
		var types = new List<Type>();
		var iterator = node.First;

		while (iterator != null)
		{
			if (iterator is IType x)
			{
				var type = x.GetType();

				if (type == Types.UNKNOWN || type.IsUnresolved)
				{
					// This operation must be aborted since type list cannot contain unresolved types
					return null;
				}
				else
				{
					types.Add(type);
				}
			}
			else
			{
				// This operation must be aborted since type list cannot contain unresolved types
				return null;
			}

			iterator = iterator.Next;
		}

		return types;
	}

	/// <summary>
	/// Tries to get the assign type from the given assign operation
	///</summary>
	private static Type? TryGetTypeFromAssignOperation(Node assign)
	{
		var operation = assign.To<OperatorNode>();

		// Try to resolve type via contextable right side of the assign operator
		if (operation.Operator == Operators.ASSIGN &&
			operation.Right is IType x)
		{
			return x.GetType();
		}

		return Types.UNKNOWN;
	}

	/// <summary>
	/// Tries to resolve the given variable by going through its references
	/// </summary>	
	public static void Resolve(Variable variable)
	{
		var types = new List<Type>();

		// Try resolving the type of the variable from its references
		foreach (var reference in variable.References)
		{
			var parent = reference.Parent;

			if (parent != null)
			{
				if (parent.GetNodeType() == NodeType.OPERATOR_NODE) // Locals
				{
					// Reference must be the destination in assign operation in order to resolve the type
					if (parent.First != reference)
					{
						continue;
					}

					var type = TryGetTypeFromAssignOperation(parent);

					if (type != Types.UNKNOWN)
					{
						types.Add(type);
					}
				}
				else if (parent.GetNodeType() == NodeType.LINK_NODE) // Members
				{
					// Reference must be the destination in assign operation in order to resolve the type
					if (parent.Last != reference)
					{
						continue;
					}

					parent = parent.Parent;

					var type = TryGetTypeFromAssignOperation(parent!);

					if (type != Types.UNKNOWN)
					{
						types.Add(type);
					}
				}
			}
		}

		// Get the shared type between the references
		var shared = Resolver.GetSharedType(types);

		if (shared != Types.UNKNOWN)
		{
			// Now the type is resolved
			variable.Type = shared;
		}
	}

	/// <summary>
	/// Tries to resolve all the variables in the given context
	/// </summary>
	private static void ResolveVariables(Context context)
	{
		foreach (var variable in context.Variables.Values)
		{
			if (variable.Type == Types.UNKNOWN)
			{
				Resolve(variable);
			}
		}

		foreach (var subcontext in context.Subcontexts)
		{
			ResolveVariables(subcontext);
		}
	}
}